modules/navier_stokes/src/problems/NavierStokesProblem.C

```diff
@@ -0,0 +1,344 @@
+
+InputParameters
+NavierStokesProblem::validParams()
+{
+  InputParameters params = FEProblem::validParams();
+  params.addClassDescription("A problem that handles Schur complement preconditioning of the "
+                             "incompressible Navier-Stokes equations");
+  params.addParam<TagName>(
+      "mass_matrix", "", "The matrix tag name corresponding to the mass matrix.");
+  params.addParam<TagName>(
+      "L_matrix",
+      "",
+      "The matrix tag name corresponding to the diffusive part of the velocity equations.");
+  params.addParam<std::vector<unsigned int>>(
+      "schur_fs_index",
+      {},
+      "if not provided then the top field split is assumed to be the "
+      "Schur split. This is a vector allow recursive nesting");
+  params.addParam<bool>("use_pressure_mass_matrix",
+                        false,
+                        "Whether to just use the pressure mass matrix as the preconditioner for "
+                        "the Schur complement");
+  params.addParam<bool>(
+      "commute_lsc",
+      false,
+      "Whether to use the commuted form of the LSC preconditioner, created by Olshanskii");
+  return params;
+}
+
+NavierStokesProblem::NavierStokesProblem(const InputParameters & parameters)
+  : FEProblem(parameters),
+    _commute_lsc(getParam<bool>("commute_lsc")),
+    _mass_matrix(getParam<TagName>("mass_matrix")),
+    _L_matrix(getParam<TagName>("L_matrix")),
+    _have_mass_matrix(!_mass_matrix.empty()),
+    _have_L_matrix(!_L_matrix.empty()),
+    _pressure_mass_matrix_as_pre(getParam<bool>("use_pressure_mass_matrix")),
+    _schur_fs_index(getParam<std::vector<unsigned int>>("schur_fs_index"))
+{
+  if (_commute_lsc)
+  {
+    if (!_have_mass_matrix)
+      paramError("mass_matrix",
+                 "A pressure mass matrix must be provided if we are commuting the LSC commutator.");
+    if (!_have_L_matrix)
+      paramError("L_matrix",
+                 "A matrix corresponding to the viscous component of the momentum equation must be "
+                 "provided if we are commuting the LSC commutator.");
+  }
+  else if (_have_L_matrix)
+    paramError("L_matrix",
+               "If not commuting the LSC commutator, then the 'L_matrix' should not be provided "
+               "because it will not be used. For Elman LSC preconditioning, L will be computed "
+               "automatically using system matrix data (e.g. the off-diagonal blocks in the "
+               "velocity-pressure system).");
+
+  if (_pressure_mass_matrix_as_pre && !_have_mass_matrix)
+    paramError("mass_matrix",
+               "If 'use_pressure_mass_matrix', then a pressure 'mass_matrix' must be provided");
+}
+
+NavierStokesProblem::~NavierStokesProblem()
+{
+  if (_Q_scale)
+    // We're destructing so don't check for errors which can throw
+    MatDestroy(&_Q_scale);
+  if (_L)
+    MatDestroy(&_L);
+}
+
+KSP
+NavierStokesProblem::findSchurKSP(KSP node, const unsigned int tree_position)
```

Comment:
why not a reference on the  KSP argument?

Reply:
Every Petsc type is actually a pointer under the hood so we are only copying a pointer